tmt/steps/discover/shell.py

```diff
@@ -13,6 +15,47 @@
     import tmt.base
 
 
+T = TypeVar('T', bound='TestDescription')
+
+
+@dataclasses.dataclass
+class TestDescription(tmt.utils.NormalizeKeysMixin, tmt.utils.SerializableContainer):
+    """
+    Keys necessary to describe a shell-based test.
+
+    Provides basic functionality for tansition between "raw" step data representation,
+    which consists of keys and values given by fmf tree and CLI options, and this
+    container representation for internal use.
+    """
+
+    name: str
+    test: str
+    # TODO: ugly circular dependency (see tmt.base.DEFAULT_TEST_DURATION_L2)
+    duration: str = '1h'
+    path: Optional[str] = None
+    environment: tmt.utils.EnvironmentType = dataclasses.field(default_factory=dict)
+
+    @classmethod
+    def from_raw(cls: Type[T], raw_data: Dict[str, Any], logger: tmt.utils.Common) -> T:
+        """
+        Unserialize step data instance from its a raw representation.
+        """
+
+        data = cls(name=raw_data['name'], test=raw_data['test'])
+        data._load_keys(raw_data, cls.__name__, logger)
+
+        return data
+
+
+@dataclasses.dataclass
+class DiscoverShellData(tmt.steps.discover.DiscoverStepData):
+    tests: List[TestDescription] = dataclasses.field(default_factory=list)
```

Comment:
Ah, thanks for pointing that out, I missed the possibility of having all those keys.

I'm not sure we could use `base.Test` here, since that one brings way more functionality, while these "smell" more like temporary containers to carry info to construct `base.Test` instances from. On paper, you're probably on the right track, if the input describes tests, we might skip the intermediate and go after `base.Test` right away, but I'd rather do that change in a separate patch. I'd *love* to do that change, it'd make the code simpler :)

Reply:
So, after some experiments, I'm afraid there's no easy way out of this - at this moment. To my understanding, base classes (`Test`, `Plan`, `Story`) were not stored in a serialized form, ever, only constructed in runtime (from fmf nodes, mappings, or other sources). `shell` discover step saved an open mapping, representing source info for future `Test` instances, and I did not realize the number of keys in this mapping is much, much larger, covering all `Test` attributes.

Re-using classes is certainly the way to go, it's definitely doable. One possibility would be to extract keys from `Test` into some kind of `_BaseTest`, add serialization support, and use it as a base class for `Test` and `shell`'s step data. Another option would be to add `SerializableContainer` as the `Test` base class directly, but `Test` is not a dataclass, therefore not compatible. Also, `Test` applies validation automagically, which is not needed for `shell` step data. There are also other possible solutions.

The issue is, that every single of these solutions is quite complicated, and would involve a fair amount of additional changes. My proposal at this point would therefore be 1. cheap & dirty copy & paste of all `Test` attributes, 2. add big fat `TODO` comment, 3. file an issue for refactoring, 4. /me preparing a patch focusing on this particular issue later. Trust me, I hate repetition, and I for one would opt for teaching base classes serialization, so they could be seamlessly used in this way, so I don't mind investing time to improving this area, I'd rather do it in a separate patch while keeping things ugly temporarily.

Reply:
+1 to quick & dirty now to have "less" changes

Reply:
Then see the diff, I pushed the changes yesterday evening. I'll file the follow-up issue to keep things tracked.

Reply:
Agreed. Looks reasonable. Let's do it this way.